tests/publiccloud/ipa.pm

```diff
@@ -11,209 +11,52 @@
-      =~ /"InstanceId":\s*"([^"]+)"/;
-    $self->{'ipa_instance_id'} = $instance_id;
-
-
-    assert_script_run(
-        "ipa test ec2 "
-          . "--access-key-id '"
-          . get_required_var('PUBLIC_CLOUD_KEY_ID') . "' "
-          . "--secret-access-key '"
-          . get_required_var('PUBLIC_CLOUD_KEY_SECRET') . "' "
-          . "-D 'IPA test $ami_id' "
-          . "--distro sles "
-          . "-R '$instance_id' "
-          . "--region '$region' "
-          . "-u ec2-user "
-          . "--ssh-private-key-file QA_SSH_KEY.pem "
-          . "--ssh-key-name '" . $ssh_key_name . "' "
-          . "--results-dir ipa_results "
-          . get_ipa_tests(),
-        timeout => 600
-    );
-    delete $self->{'ipa_instance_id'};
-
-    ec2_cleanup;
-}
+    select_virtio_console();
 
-sub az_run_ipa {
-    my ($self) = @_;
+    my $provider = $self->{provider} = $self->provider_factory();
 
-    my $clientid      = get_required_var('PUBLIC_CLOUD_KEY_ID');
-    my $clientsecret  = get_required_var('PUBLIC_CLOUD_KEY_SECRET');
-    my $subscription  = get_required_var('PUBLIC_CLOUD_SUBSCRIPTION_ID');
-    my $tenantid      = get_required_var('PUBLIC_CLOUD_TENANT_ID');
-    my $image         = get_required_var('PUBLIC_CLOUD_IMAGE_ID');
-    my $instance_type = get_var('PUBLIC_CLOUD_INSTANCE_TYPE', 'Standard_A2');
-    my $region        = get_var('PUBLIC_CLOUD_REGION', 'westeurope');
-
-    my $credentials = <<EOT;
-{
-  "clientId": "$clientid", 
-  "clientSecret": "$clientsecret", 
-  "subscriptionId": "$subscription", 
-  "tenantId": "$tenantid", 
-  "activeDirectoryEndpointUrl": "https://login.microsoftonline.com", 
-  "resourceManagerEndpointUrl": "https://management.azure.com/", 
-  "activeDirectoryGraphResourceId": "https://graph.windows.net/", 
-  "sqlManagementEndpointUrl": "https://management.core.windows.net:8443/", 
-  "galleryEndpointUrl": "https://gallery.azure.com/", 
-  "managementEndpointUrl": "https://management.core.windows.net/"
-}
-EOT
-
-    save_tmp_file("azure_credentials.txt", $credentials);
-    assert_script_run('curl -O ' . autoinst_url . "/files/azure_credentials.txt");
-
-    assert_script_run('ssh-keygen -b 2048 -t rsa -q -N "" -f ~/.ssh/id_rsa');
-
-    assert_script_run(
-        "ipa test azure "
-          . "--service-account-file azure_credentials.txt "
-          . "-D 'IPA test $image' "
-          . "--distro sles "
-          . "--ssh-private-key-file ~/.ssh/id_rsa "
-          . "--region '$region' "
-          . "--results-dir ipa_results "
-          . "--cleanup "
-          . "--instance-type '$instance_type' "
-          . "-i '$image' "
-          . get_ipa_tests(),
-        timeout => 1200
+    my $ipa = $provider->ipa(
+        instance_type => get_var('PUBLIC_CLOUD_INSTANCE_TYPE'),
+        cleanup       => 1,
+        image_id      => $self->get_image_id($provider),
+        tests         => get_required_var('PUBLIC_CLOUD_IPA_TESTS'),
+        results_dir   => 'ipa_results'
```

Comment:
I think region is common for all 3 providers, would it make sense to also add region here and set it up as a openqa job variable? 

    region => get_required_var('PUBLIC_CLOUD_REGION'),

Reply:
Good point.
Currently the region is set during the constructor, see `lib/publiccloud/basetest.pm` and then used as member variable from provider.
If it is ok for you, I would keep it this way and do not change the region during test.